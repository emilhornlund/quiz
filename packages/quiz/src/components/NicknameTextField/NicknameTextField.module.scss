@use '../../styles/colors';
@use '../../styles/helpers';
@use '../../styles/variables';

.nicknameTextFieldContainer {
  display: flex;
  flex-direction: row;
  width: 100%;

  @include helpers.mobile {
    column-gap: variables.$mobile-narrow-layout-gutter
  }

  @include helpers.tablet {
    column-gap: variables.$tablet-narrow-layout-gutter
  }

  @include helpers.desktop {
    column-gap: variables.$desktop-narrow-layout-gutter;
  }

  > :first-child {
    flex: 1;
  }
}