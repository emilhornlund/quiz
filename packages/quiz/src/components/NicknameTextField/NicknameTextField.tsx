import { faRefresh } from '@fortawesome/free-solid-svg-icons'
import {
  generateNickname,
  PLAYER_NICKNAME_MAX_LENGTH,
  PLAYER_NICKNAME_MIN_LENGTH,
  PLAYER_NICKNAME_REGEX,
} from '@quiz/common'
import React, { FC } from 'react'

import Button from '../Button'
import { TextField } from '../index.ts'

import styles from './NicknameTextField.module.scss'

export interface NicknameTextFieldProps {
  value?: string
  placeholder?: string
  disabled?: boolean
  onChange?: (value: string) => void
  onValid?: (valid: boolean) => void
}

const NicknameTextField: FC<NicknameTextFieldProps> = ({
  value,
  placeholder = 'Default Nickname',
  disabled = false,
  onChange,
  onValid,
}) => {
  const handleRefreshNicknameClick = () => {
    onChange?.(generateNickname())
  }

  return (
    <div className={styles.nicknameTextFieldContainer}>
      <TextField
        id="default-nickname-textfield"
        type="text"
        kind="primary"
        placeholder={placeholder}
        value={value}
        minLength={PLAYER_NICKNAME_MIN_LENGTH}
        maxLength={PLAYER_NICKNAME_MAX_LENGTH}
        regex={PLAYER_NICKNAME_REGEX}
        disabled={disabled}
        onChange={(value) => onChange?.(value as string)}
        onValid={(valid) => onValid?.(valid)}
        required
      />
      <Button
        id="shuffle-nickname-button"
        type="button"
        kind="primary"
        disabled={disabled}
        icon={faRefresh}
        onClick={handleRefreshNicknameClick}
      />
    </div>
  )
}

export default NicknameTextField
