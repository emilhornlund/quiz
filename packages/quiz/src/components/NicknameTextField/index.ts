export type { NicknameTextFieldProps } from './NicknameTextField'
export { default } from './NicknameTextField'
