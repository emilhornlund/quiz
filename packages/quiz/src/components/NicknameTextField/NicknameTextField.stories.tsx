import type { Meta, StoryObj } from '@storybook/react'

import NicknameTextField from './NicknameTextField'

const meta = {
  title: 'Inputs/NicknameTextField',
  component: NicknameTextField,
  tags: ['autodocs'],
} satisfies Meta<typeof NicknameTextField>

export default meta
type Story = StoryObj<typeof meta>

export const Default = {
  args: {
    value: 'FrostyBear',
    disabled: false,
    onChange: () => undefined,
    onValid: () => undefined,
  },
} satisfies Story

export const Disabled = {
  args: {
    value: 'FrostyBear',
    disabled: true,
    onChange: () => undefined,
    onValid: () => undefined,
  },
} satisfies Story
