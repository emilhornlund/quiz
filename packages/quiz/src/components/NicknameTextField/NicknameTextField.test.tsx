import { render } from '@testing-library/react'
import React from 'react'
import { describe, expect, test } from 'vitest'

import NicknameTextField from './NicknameTextField'

describe('NicknameTextField', () => {
  test('should render NicknameTextField with default props', async () => {
    const { container } = render(<NicknameTextField value="FrostyBear" />)

    expect(container).toMatchSnapshot()
  })
})
