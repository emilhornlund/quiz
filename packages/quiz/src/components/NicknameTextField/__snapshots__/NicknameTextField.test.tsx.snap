// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NicknameTextField > should render NicknameTextField with default props 1`] = `
<div>
  <div
    class="nicknameTextFieldContainer"
  >
    <div
      class="inputContainer"
    >
      <div
        class="textFieldInputContainer textFieldInputKindPrimary"
      >
        <input
          class="textfield"
          data-testid="test-default-nickname-textfield-textfield"
          id="default-nickname-textfield"
          maxlength="20"
          minlength="2"
          name="default-nickname-textfield"
          placeholder="Default Nickname"
          type="text"
          value="FrostyBear"
        />
      </div>
    </div>
    <div
      class="buttonInputContainer buttonInputKindPrimary"
    >
      <button
        data-testid="test-shuffle-nickname-button-button"
        id="shuffle-nickname-button"
        name="shuffle-nickname-button"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-arrows-rotate"
          data-icon="arrows-rotate"
          data-prefix="fas"
          role="img"
          viewBox="0 0 512 512"
        >
          <path
            d="M65.9 228.5c13.3-93 93.4-164.5 190.1-164.5 53 0 101 21.5 135.8 56.2 .2 .2 .4 .4 .6 .6l7.6 7.2-47.9 0c-17.7 0-32 14.3-32 32s14.3 32 32 32l128 0c17.7 0 32-14.3 32-32l0-128c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 53.4-11.3-10.7C390.5 28.6 326.5 0 256 0 127 0 20.3 95.4 2.6 219.5 .1 237 12.2 253.2 29.7 255.7s33.7-9.7 36.2-27.1zm443.5 64c2.5-17.5-9.7-33.7-27.1-36.2s-33.7 9.7-36.2 27.1c-13.3 93-93.4 164.5-190.1 164.5-53 0-101-21.5-135.8-56.2-.2-.2-.4-.4-.6-.6l-7.6-7.2 47.9 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L32 320c-8.5 0-16.7 3.4-22.7 9.5S-.1 343.7 0 352.3l1 127c.1 17.7 14.6 31.9 32.3 31.7S65.2 496.4 65 478.7l-.4-51.5 10.7 10.1c46.3 46.1 110.2 74.7 180.7 74.7 129 0 235.7-95.4 253.4-219.5z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
